sklearn/utils/_array_api.py

```diff
@@ -74,22 +88,36 @@ def _check_array_api_dispatch(array_api_dispatch):
             )
 
 
-def device(x):
+def device(*array_list):
     """Hardware device the array data resides on.
 
     Parameters
     ----------
-    x : array
-        Array instance from NumPy or an array API compatible library.
+    *array_list : arrays
+        List of array instances from NumPy or an array API compatible library.
 
     Returns
     -------
     out : device
         `device` object (see the "Device Support" section of the array API spec).
     """
-    if isinstance(x, (numpy.ndarray, numpy.generic)):
-        return "cpu"
-    return x.device
+    if not array_list:
+        raise ValueError("At least one input array expected, got none.")
+
+    devices = set()
+    for array in array_list:
+        if isinstance(array, (numpy.ndarray, numpy.generic)) or not hasattr(
+            array, "device"
+        ):
+            devices.add("cpu")
+        else:
+            devices.add(array.device)
```

Comment:
Actually this line can trigger `TypeError: unhashable type: 'cupy.cuda.device.Device'`.

Maybe we can do a for a loop on a `unique_devices` list and test for list member ship instead of set membership to only rely on equality checks instead of hash + equality checks:

```python
>>> import cupy.array_api as xp
>>> xp.asarray([0]).device in [xp.asarray([0]).device]
True
>>> xp.asarray([0]).device in {xp.asarray([0]).device}
Traceback (most recent call last):
  Cell In[12], line 1
    xp.asarray([0]).device in {xp.asarray([0]).device}
TypeError: unhashable type: 'cupy.cuda.device.Device'
```